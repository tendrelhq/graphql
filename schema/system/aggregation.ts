/** @gqlType */
export type Aggregate = {
  /**
   * The group, or bucket, that uniquely identifies this aggregate.
   * For example, this will be one of the `groupByTag`s passed to `trackingAgg`.
   *
   * @gqlField
   */
  group: string;

  /**
   * The computed aggregate value.
   *
   * Currently, this will always be a string value representing a duration in
   * seconds, e.g. "360" -> 360 seconds. `null` will be returned when no such
   * aggregate can be computed, e.g. "time in planned downtime" when no "planned
   * downtime" events exist.
   *
   * @gqlField
   */
  value: string | null;
};
