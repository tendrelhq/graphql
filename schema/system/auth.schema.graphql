interface Identity implements Component {
  id: Entity!
}
