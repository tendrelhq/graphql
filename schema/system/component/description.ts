import type { ID } from "grats";
import type { Component } from "../component";
import type { Refetchable } from "../node";
import type { Context } from "@/schema/types";
import { normalizeBase64 } from "@/util";
import { decodeGlobalId } from "..";
import type { DynamicString, DynamicStringInput } from "../i18n";

export type ConstructorArgs = {
  id: ID;
};

/**
 * @gqlType
 */
export class Description implements Component, Refetchable {
  readonly __typename = "Description" as const;
  readonly _type: string;
  readonly _id: string;
  readonly id: ID;

  constructor(args: ConstructorArgs) {
    this.id = normalizeBase64(args.id);
    const { type, id } = decodeGlobalId(this.id);
    this._type = type;
    this._id = id;
  }

  /**
   * @gqlField
   * @deprecated Use Description.locale and/or Description.value.
   */
  async description(ctx: Context): Promise<DynamicString> {
    return await ctx.orm.dynamicString.load(this.id);
  }

  /**
   * @gqlField
   */
  async locale(ctx: Context): Promise<string> {
    const s = await ctx.orm.dynamicString.load(this.id);
    return s.locale;
  }

  /**
   * @gqlField
   */
  async value(ctx: Context): Promise<string> {
    const s = await ctx.orm.dynamicString.load(this.id);
    return s.value;
  }
}

/**
 * @gqlInput
 */
export type DescriptionInput = {
  id?: ID | null;
  value: DynamicStringInput;
};
