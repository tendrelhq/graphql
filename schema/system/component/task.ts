import { sql } from "@/datasources/postgres";
import type { Trackable } from "@/schema/platform/tracking";
import type { Mutation } from "@/schema/root";
import type { Component } from "@/schema/system/component";
import type { Overridable } from "@/schema/system/overridable";
import type { Timestamp } from "@/schema/system/temporal";
import type { Context } from "@/schema/types";
import { assertNonNull } from "@/util";
import { GraphQLError } from "graphql";
import type { ID } from "grats";
import { match } from "ts-pattern";
import { decodeGlobalId } from "..";
import type { Refetchable } from "../node";
import type { Connection } from "../pagination";
import type { DisplayName, Named } from "./name";

export type ConstructorArgs = {
  id: ID;
};

/**
 * A system-level component that identifies an Entity as being applicable to
 * tendrel's internal "task processing pipeline". In practice, Tasks most often
 * represent "jobs" performed by humans. However, this need not always be the
 * case.
 *
 * Technically speaking, a Task represents a (1) *named asynchronous process*
 * that (2) exists in one of three states: open, in progress, or closed.
 *
 * @gqlType
 */
export class Task implements Component, Named, Refetchable, Trackable {
  readonly __typename = "Task" as const;
  readonly _type: string;
  readonly _id: string;
  readonly id: ID;

  constructor(
    args: ConstructorArgs,
    private ctx: Context,
  ) {
    // Note that Postgres will sometimes add newlines when we `encode(...)`.
    this.id = args.id.replace(/\n/g, "");
    // Private.
    const { type, id } = decodeGlobalId(this.id);
    this._type = type;
    this._id = id;
  }

  /** @gqlField */
  async displayName(): Promise<DisplayName> {
    return await this.ctx.orm.displayName.load(this.id);
  }

  async root(): Promise<Task | null> {
    if (this._type !== "workinstance") return null;

    const [row] = await sql<[{ id: ID }?]>`
      SELECT encode(('workinstance:' || og.id)::bytea, 'base64') AS id
      FROM public.workinstance AS wi
      INNER JOIN public.workinstance AS og
          ON wi.workinstanceoriginatorid = og.workinstanceid
      WHERE
          wi.id = ${this._id}
          AND wi.workinstanceid IS DISTINCT FROM wi.workinstanceoriginatorid;
    `;

    // If there is no row, there is no originator. This implies that this Task
    // is the root of the chain, i.e. the originator.
    if (!row) return this;

    return new Task({ id: row.id }, this.ctx);
  }

  /** @gqlField */
  async state(): Promise<TaskState | null> {
    // Only workinstances have statuses.
    if (this._type !== "workinstance") return null;

    const [row] = await sql<
      [
        {
          status: string;
          create_date: string;
          target_start_date?: string;
          ov_target_start_date?: string;
          start_date?: string;
          ov_start_date?: string;
          close_date?: string;
          ov_close_date?: string;
          time_zone: string;
        }?,
      ]
    >`
      SELECT
          wis.systagtype AS status,
          wi.workinstancecreateddate AS create_date,
          wi.workinstancetargetstartdate AS target_start_date,
          nullif(ov_target.workresultinstancevalue, '')::timestamptz AS ov_target_start_date,
          wi.workinstancestartdate AS start_date,
          nullif(ov_start.workresultinstancevalue, '')::timestamptz AS ov_start_date,
          wi.workinstancecompleteddate AS close_date,
          nullif(ov_close.workresultinstancevalue, '')::timestamptz AS ov_close_date,
          wi.workinstancetimezone AS time_zone
      FROM public.workinstance AS wi
      INNER JOIN public.systag AS wis
          ON wi.workinstancestatusid = wis.systagid
      LEFT JOIN public.workresult AS ov_target_f
          ON
              wi.workinstanceworktemplateid = ov_target_f.workresultworktemplateid
              AND ov_target_f.workresultisprimary = true
              AND ov_target_f.workresulttypeid IN (
                  SELECT systagid
                  FROM public.systag
                  WHERE systagtype = 'Override Target Start Time'
              )
      LEFT JOIN public.workresultinstance AS ov_target
          ON
              wi.workinstanceid = ov_target.workresultinstanceworkinstanceid
              AND ov_target_f.workresultid = ov_target.workresultinstanceworkresultid
      LEFT JOIN public.workresult AS ov_start_f
          ON
              wi.workinstanceworktemplateid = ov_start_f.workresultworktemplateid
              AND ov_start_f.workresultisprimary = true
              AND ov_start_f.workresulttypeid IN (
                  SELECT systagid
                  FROM public.systag
                  WHERE systagtype = 'Override Start Time'
              )
      LEFT JOIN public.workresultinstance AS ov_start
          ON
              wi.workinstanceid = ov_start.workresultinstanceworkinstanceid
              AND ov_start_f.workresultid = ov_start.workresultinstanceworkresultid
      LEFT JOIN public.workresult AS ov_close_f
          ON
              wi.workinstanceworktemplateid = ov_close_f.workresultworktemplateid
              AND ov_close_f.workresultisprimary = true
              AND ov_close_f.workresulttypeid IN (
                  SELECT systagid
                  FROM public.systag
                  WHERE systagtype = 'Override Start Time'
              )
      LEFT JOIN public.workresultinstance AS ov_close
          ON
              wi.workinstanceid = ov_close.workresultinstanceworkinstanceid
              AND ov_close_f.workresultid = ov_close.workresultinstanceworkresultid
      WHERE wi.id = ${this._id};
    `;

    if (!row) {
      console.warn(`No TaskState for ${this}... which is odd?`);
      return null;
    }

    return match(row.status)
      .with(
        "Open",
        () =>
          ({
            __typename: "Open" as const,
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
          }) satisfies Open,
      )
      .with(
        "In Progress",
        () =>
          ({
            __typename: "InProgress",
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
            inProgressAt: {
              override:
                row.start_date && row.ov_start_date
                  ? {
                      previousValue: {
                        epochMilliseconds: row.start_date,
                        timeZone: row.time_zone,
                      },
                    }
                  : null,
              value: {
                epochMilliseconds: assertNonNull(
                  row.ov_start_date ?? row.start_date,
                  `Task ${this}, in state 'In Progress', has no start date`,
                ),
                timeZone: row.time_zone,
              },
            },
          }) satisfies InProgress,
      )
      .with(
        "Cancelled",
        () =>
          ({
            __typename: "Closed",
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
            inProgressAt: row.start_date
              ? {
                  override: row.ov_start_date
                    ? {
                        previousValue: {
                          epochMilliseconds: row.start_date,
                          timeZone: row.time_zone,
                        },
                      }
                    : null,
                  value: {
                    epochMilliseconds: row.ov_start_date ?? row.start_date,
                    timeZone: row.time_zone,
                  },
                }
              : null,
            closedAt: {
              override:
                row.close_date && row.ov_close_date
                  ? {
                      previousValue: {
                        epochMilliseconds: row.close_date,
                        timeZone: row.time_zone,
                      },
                    }
                  : null,
              value: {
                epochMilliseconds: assertNonNull(
                  row.ov_close_date ?? row.close_date,
                  `Task ${this}, in state 'Cancelled', has no close date`,
                ),
                timeZone: row.time_zone,
              },
            },
          }) satisfies Closed,
      )
      .with(
        "Complete",
        () =>
          ({
            __typename: "Closed",
            openedAt: {
              override: null,
              value: {
                epochMilliseconds: row.create_date,
                timeZone: row.time_zone,
              },
            },
            inProgressAt: row.start_date
              ? {
                  override: row.ov_start_date
                    ? {
                        previousValue: {
                          epochMilliseconds: row.start_date,
                          timeZone: row.time_zone,
                        },
                      }
                    : null,
                  value: {
                    epochMilliseconds: row.ov_start_date ?? row.start_date,
                    timeZone: row.time_zone,
                  },
                }
              : null,
            closedAt: {
              override:
                row.close_date && row.ov_close_date
                  ? {
                      previousValue: {
                        epochMilliseconds: row.close_date,
                        timeZone: row.time_zone,
                      },
                    }
                  : null,
              value: {
                epochMilliseconds: assertNonNull(
                  row.ov_close_date ?? row.close_date,
                  `Task ${this}, in state 'Closed', has no close date`,
                ),
                timeZone: row.time_zone,
              },
            },
          }) satisfies Closed,
      )
      .otherwise(s => {
        console.warn(`Unknown TaskState type '${s}'`);
        return null;
      });
  }

  /**
   * Entrypoint into the "tracking system(s)" for the given Task.
   * At the moment, sub-task tracking is not supported and therefore `null` will
   * always be returned for this field.
   *
   * @gqlField
   */
  async tracking(): Promise<Connection<Trackable> | null> {
    return null;
  }

  toString() {
    return `'${this.id}' (${this._type}:${this._id})`;
  }
}

// FIXME: this might be a confusing name in combination with TaskStateMachine...
// Probably should just name it TaskStatus?
/** @gqlUnion */
export type TaskState = Open | InProgress | Closed;

/** @gqlType */
export type Open = {
  __typename: "Open";

  /** @gqlField */
  dueAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  openedAt: Overridable<Timestamp>;
  /** @gqlField */
  openedBy?: string;
};

/** @gqlType */
export type InProgress = {
  __typename: "InProgress";

  /** @gqlField */
  dueAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  openedAt: Overridable<Timestamp>;
  /** @gqlField */
  openedBy?: string | null;
  /** @gqlField */
  inProgressAt: Overridable<Timestamp>;
  /** @gqlField */
  inProgressBy?: string | null;
};

/** @gqlType */
export type Closed = {
  __typename: "Closed";

  /** @gqlField */
  dueAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  openedAt: Overridable<Timestamp>;
  /** @gqlField */
  openedBy?: string | null;
  /** @gqlField */
  inProgressAt?: Overridable<Timestamp> | null;
  /** @gqlField */
  inProgressBy?: string | null;
  /** @gqlField */
  closedAt: Overridable<Timestamp>;
  /** @gqlField */
  closedBecause?: string | null;
  /** @gqlField */
  closedBy?: string;
};

/**
 * Similar to task_fsm's advance method, this method advances a Task through its
 * internal state machine. A Task's state machine has its finite set defined by
 * the variants of the {@link TaskState} union, with a dag of the form:
 * ```
 * Open -> InProgress -> Closed
 * ```
 */
export async function advance(t: Task): Promise<Task> {
  if (t._type !== "workinstance") {
    // Punt on this for now. We can come back to it.
    // This is, at least at present, used solely by the `advance` implementation
    // for StateMachine<Task>.
    throw "not yet implemented - lazy instantiation of a Task";
  }

  // As before, we assume this is the "stop task" flow.
  await sql`
    UPDATE public.workinstance
    SET
        workinstancestatusid = 710,
        workinstancecompleteddate = now(),
        workinstancemodifieddate = now()
        -- TODO: workinstancemodifiedby
    WHERE id = ${t._id};
  `;

  return t;
}
