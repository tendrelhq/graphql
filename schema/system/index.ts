import z, { type Infer } from "myzod";

const GlobalId = z.string().map(id => {
  const parts = id.split(":");
  if (parts.length !== 2) {
    throw "invariant violated: invalid global identifier";
  }
  return {
    type: parts[0],
    id: parts[1],
  };
});

/**
 * Global Identifier. Externally it just an opaque string. Internally it may
 * have some meaning. Right now this is just the base64 encoding of the
 * object's underlying type (e.g. workinstance) and its uuid (i.e. primary
 * key).
 */
type GlobalId = Infer<typeof GlobalId>;

export function decodeGlobalId(id: unknown): GlobalId {
  if (typeof id !== "string") {
    throw "invariant violated: global ids should be string";
  }
  return GlobalId.parse(
    Buffer.from(decodeURIComponent(id), "base64").toString(),
  );
}

export function encodeGlobalId({ type, id }: GlobalId) {
  return Buffer.from(`${type}:${id}`).toString("base64");
}

// biome-ignore lint/suspicious/noExplicitAny:
export function isGlobalId(value: any): value is GlobalId {
  if (typeof value?.type === "string" && typeof value?.id === "string") {
    return true;
  }

  return false;
}
