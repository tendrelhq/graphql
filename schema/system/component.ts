import { sql } from "@/datasources/postgres";
import { assert, assertNonNull, buildPaginationArgs, mapOrElse } from "@/util";
import { GraphQLError } from "graphql";
import type { ID, Int } from "grats";
import type { Fragment } from "postgres";
import { match } from "ts-pattern";
import { decodeGlobalId } from ".";
import {
  Attachment,
  type ConstructorArgs as AttachmentConstructorArgs,
} from "../platform/attachment";
import type { Context } from "../types";
import type { Description } from "./component/description";
import type { DisplayName } from "./component/name";
import { type ConstructorArgs, Task } from "./component/task";
import type { Connection, Edge, PageInfo } from "./pagination";
import type { Timestamp } from "./temporal";

/**
 * Components characterize Entities as possessing a particular trait.
 * They are just simple structs, holding all data necessary to model that trait.
 *
 * @gqlInterface
 */
export interface Component {
  readonly __typename: string;

  /**
   * @gqlField
   * @killsParentOnException
   */
  readonly id: ID;
}

export const field$fragment: Fragment = sql`
select
    f.id,
    case
        when f.type = 'Boolean'
        then
            jsonb_build_object(
                '__typename', 'BooleanValue', 'boolean', f.value::boolean
            )
        when f.type = 'Date'
        then
            jsonb_build_object(
                '__typename', 'TimestampValue', 'timestamp', to_timestamp(f.value::bigint / 1000.0)
            )
        when f.type = 'Number'
        then
            jsonb_build_object(
                '__typename', 'NumberValue', 'number', f.value::numeric
            )
        when f.type = 'String'
        then
            jsonb_build_object(
                '__typename', 'StringValue', 'string', f.value
            )
        else '{}'::jsonb
    end as value,
    case
        when f.type = 'Boolean' then 'boolean'
        when f.type = 'Date' then 'timestamp'
        when f.type = 'Number' then 'number'
        when f.type = 'String' then 'string'
        else 'unknown'
    end as "valueType"
from field as f
`;

/** @gqlType */
export type Field = {
  /**
   * Unique identifier for this Field.
   *
   * @gqlField
   */
  id: ID;

  /**
   * The value for this Field, if any. This field will always be present (when
   * requested) for the given Field so as to convey the underlying data type of
   * the (raw data) value. The underlying (raw data) value can be `null`.
   *
   * @gqlField
   */
  value: Value;

  /**
   * The type of data underlying `value`. This is provided as a convenience when
   * interacting with field-level edits through other apis.
   *
   * @gqlField
   */
  valueType: ValueType;
};

/** @gqlType */
export type ValueCompletion = {
  /** @gqlField */
  value: Value;
};

/**
 * Intended to provide "auto-completion" in a frontend setting, this API returns
 * *distinct known values* for a given Field. For Fields without constraints
 * (which is most of them), this will return a "frecency" list of previously
 * used values for the given Field. When constraints are involved, the
 * completion list represents the *allowed* set of values for the given Field.
 *
 * Note that "frecency" is not currently implemented. For such Fields (i.e. those
 * without constraints) you will simply get back an empty completion list.
 *
 * Note also that currently there is no enforcement of the latter, constraint-based
 * semantic in the backend! The client *must* validate user input using the
 * completion list *before* issuing, for example, an `applyFieldEdits` mutation.
 * Otherwise the backend will gladly accept arbitrary values (assuming they are,
 * of course, of the correct type).
 *
 * Note also that pagination is not currently implemented.
 *
 * @gqlField
 */
export async function completions(
  f: Field,
  ctx: Context,
): Promise<Connection<ValueCompletion>> {
  const { type, id, suffix } = decodeGlobalId(f.id);
  // FIXME: we aren't using worktemplateconstraintconstrainedtypeid here.
  // This feels like a gap. Although in the ideal model result type would inform
  // our choice of algorithm: frecency vs constraint.
  // TODO: localization.
  const cte = match(type)
    .with(
      "workresult",
      () => sql`
        select
          ''::text as id, -- unused, purely for fragment reuse
          s.systagtype as "type",
          c.custagtype as "value"
        from public.workresult as wr
        inner join public.worktemplate as wt
          on wr.workresultworktemplateid = wt.worktemplateid
        inner join public.systag as s
          on wr.workresulttypeid = s.systagid
        inner join public.worktemplateconstraint as wtc
          on wt.id = wtc.worktemplateconstrainttemplateid
          and wr.id = wtc.worktemplateconstraintresultid
        inner join public.custag as c
          on wtc.worktemplateconstraintconstraintid = c.custaguuid
          and (c.custagenddate is null or c.custagenddate > now())
        where wr.id = ${id}
        order by c.custagorder asc
      `,
    )
    .with("workresultinstance", () => {
      const instance = id; // workinstanceuuid
      const field = assertNonNull(suffix?.at(0), "invalid global identifier"); // workresultuuid
      return sql`
        select
          ''::text as id, -- unused, purely for fragment reuse
          s.systagtype as "type",
          c.custagtype as "value"
        from public.workinstance as wi
        inner join public.worktemplate as wt
          on wi.workinstanceworktemplateid = wt.worktemplateid
        inner join public.workresult as wr
          on wt.worktemplateid = wr.workresultworktemplateid
          and wr.id = ${field}
        inner join public.systag as s
          on wr.workresulttypeid = s.systagid
        inner join public.worktemplateconstraint as wtc
          on wt.id = wtc.worktemplateconstrainttemplateid
          and wr.id = wtc.worktemplateconstraintresultid
        inner join public.custag as c
          on wtc.worktemplateconstraintconstraintid = c.custaguuid
          and (c.custagenddate is null or c.custagenddate > now())
        where wi.id = ${instance}
        order by c.custagorder asc
      `;
    })
    .otherwise(() => {
      throw "";
    });

  const rows = await sql<ValueCompletion[]>`
    with field as (${cte})
    ${field$fragment}
  `;

  return {
    edges: rows.map(row => ({ node: row, cursor: "" })),
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    totalCount: rows.length,
  };
}

/**
 * Field attachments.
 *
 * @gqlField
 */
export async function attachments(
  f: Field,
  ctx: Context,
  args: {
    first?: Int | null;
    last?: Int | null;
    before?: string | null;
    after?: string | null;
  },
): Promise<Connection<Attachment>> {
  const { type, id, suffix } = decodeGlobalId(f.id);
  // Only instances can have attachments.
  if (type !== "workresultinstance") {
    return {
      edges: [],
      pageInfo: {
        hasNextPage: false,
        hasPreviousPage: false,
      },
      totalCount: 0,
    };
  }

  const p = buildPaginationArgs(args, {
    defaultLimit: ctx.limits.fieldAttachmentPaginationDefaultLimit,
    maxLimit: ctx.limits.fieldAttachmentPaginationMaxLimit,
  });
  const rows = await sql<AttachmentConstructorArgs[]>`
    select
        encode(('workpictureinstance:' || a.workpictureinstanceuuid)::bytea, 'base64') as id,
        a.workpictureinstancestoragelocation as url
    from public.workresultinstance as field
    inner join public.workpictureinstance as a
        on field.workresultinstanceworkinstanceid = a.workpictureinstanceworkinstanceid
        and field.workresultinstanceid = a.workpictureinstanceworkresultinstanceid
    where
        field.workresultinstanceworkinstanceid = (
            select workinstanceid
            from public.workinstance
            where id = ${id}
        )
        and field.workresultinstanceworkresultid = (
            select workresultid
            from public.workresult
            where id = ${assertNonNull(suffix?.at(0), "invariant violated")}
        )
        ${mapOrElse(
          p.cursor,
          cursor => sql`
        and
            (a.workpictureinstancemodifieddate, a.workpictureinstanceid)
            ${p.direction === "forward" ? sql`<` : sql`>`}
            (
                select c.workpictureinstancemodifieddate, c.workpictureinstanceid
                from public.workpictureinstance as c
                where c.workpictureinstanceuuid = ${cursor.id}
            )
          `,
          sql``,
        )}
    order by a.workpictureinstancemodifieddate desc, a.workpictureinstanceid desc
    limit ${p.limit + 1};
  `;

  const n1 = rows.length > p.limit ? rows.pop() : undefined;
  const edges: Edge<Attachment>[] = rows.map(row => ({
    cursor: row.id,
    node: new Attachment(row, ctx),
  }));
  const pageInfo: PageInfo = {
    startCursor: edges.at(0)?.cursor,
    endCursor: edges.at(-1)?.cursor,
    hasNextPage: p.direction === "forward" && !!n1,
    hasPreviousPage: p.direction === "backward" && !!n1,
  };

  const [{ count }] = await sql<[{ count: bigint }]>`
    select count(*)
    from public.workresultinstance as field
    inner join public.workpictureinstance as a
        on field.workresultinstanceworkinstanceid = a.workpictureinstanceworkinstanceid
        and field.workresultinstanceid = a.workpictureinstanceworkresultinstanceid
    where
        field.workresultinstanceworkinstanceid = (
            select workinstanceid
            from public.workinstance
            where id = ${id}
        )
        and field.workresultinstanceworkresultid = (
            select workresultid
            from public.workresult
            where id = ${assertNonNull(suffix?.at(0), "invariant violated")}
        )
  `;

  return {
    edges,
    pageInfo,
    totalCount: Number(count),
  };
}

/**
 * Description of a Field.
 *
 * @gqlField
 */
export async function description(
  field: Field,
  ctx: Context,
): Promise<Description | null> {
  return await ctx.orm.description.load(field.id);
}

/**
 * Display name for a Field.
 *
 * @gqlField
 */
export async function name(field: Field, ctx: Context): Promise<DisplayName> {
  return await ctx.orm.displayName.load(field.id);
}

/** @gqlField */
export async function parent(field: Field): Promise<Task> {
  const { type, id } = decodeGlobalId(field.id);
  const [row] = await match(type)
    .with(
      "workresult",
      () => sql<[ConstructorArgs]>`
        select engine1.base64_encode(convert_to('worktemplate:' || id, 'utf8')) as id
        from public.worktemplate
        where worktemplateid in (
          select workresultworktemplateid
          from public.workresult
          where id = ${id}
        )
      `,
    )
    .with(
      "workresultinstance",
      () => sql<[ConstructorArgs]>`
        select engine1.base64_encode(convert_to('workinstance:' || id, 'utf8')) as id
        from public.workinstance
        where id = ${id}
      `,
    )
    .otherwise(() => {
      throw "invariant violated";
    });

  return new Task(row);
}

export type FieldDefinitionInput = {
  name: string;
  type: string;
  description?: string | null;
  isDraft?: boolean | null;
  isPrimary?: boolean | null;
  order?: number | null;
  referenceType?: string | null;
  value?: ValueInput | null;
  widget?: string | null;
};

/** @gqlInput */
export type FieldInput = {
  field: ID;
  value?: ValueInput | null;
  /**
   * Must match the type of the `value`, e.g.
   * ```typescript
   * if (field.valueType === "string") {
   *   assert(field.value === null || "string" in field.value);
   * }
   * ```
   */
  valueType: ValueType;
};

/** @gqlEnum */
export type ValueType =
  | "boolean"
  | "entity"
  | "number"
  | "string"
  | "timestamp"
  /** For backwards compatibility. */
  | "unknown";

/** @gqlUnion */
export type Value =
  | BooleanValue
  | EntityValue
  | NumberValue
  | StringValue
  | TimestampValue;

/** @gqlType */
type BooleanValue = {
  __typename: "BooleanValue";

  /** @gqlField */
  boolean: boolean | null;
};

/** @gqlType */
type EntityValue = {
  __typename: "EntityValue";

  /** @gqlField */
  entity: Component | null;
};

/** @gqlType */
type NumberValue = {
  __typename: "NumberValue";

  /** @gqlField */
  number: Int | null;
};

/** @gqlType */
type StringValue = {
  __typename: "StringValue";

  /** @gqlField */
  string: string | null;
};

/** @gqlType */
type TimestampValue = {
  __typename: "TimestampValue";

  /** @gqlField */
  timestamp: Timestamp | null;
};

/**
 * @gqlInput
 * @oneOf
 */
export type ValueInput =
  // Boolean
  | {
      boolean: boolean;
    }
  // Entity
  | {
      id: ID;
    }
  // Number
  | {
      number: Int;
    }
  // String
  | {
      string: string;
    }
  // Timestamp
  | {
      /**
       * ISO 8601 format.
       */
      timestamp: string;
    };
