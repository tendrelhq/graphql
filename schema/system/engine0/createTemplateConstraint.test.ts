import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { sql } from "@/datasources/postgres";
import { schema } from "@/schema/final";
import { Location } from "@/schema/platform/archetype/location";
import { decodeGlobalId } from "@/schema/system";
import { Task } from "@/schema/system/component/task";
import {
  createTestContext,
  execute,
  findAndEncode,
  getFieldByName,
} from "@/test/prelude";
import { map } from "@/util";
import { TestCreateTemplateConstraintDocument } from "./createTemplateConstraint.test.generated";

const ctx = await createTestContext();

describe("createTemplateConstraint", () => {
  let CUSTOMER: string;
  let TEMPLATE: Task;
  let LOCATION: Location;

  test("create a new constraint", async () => {
    const result = await execute(schema, TestCreateTemplateConstraintDocument, {
      template: TEMPLATE.id,
      location: LOCATION.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data?.createTemplateConstraint).toMatchObject({
      constraint: {
        id: expect.any(String),
      },
      diagnostics: [],
      instantiations: [],
    });

    const [{ count }] = await sql`
      select count(*)
      from public.worktemplateconstraint
      where worktemplateconstraintcustomeruuid = ${decodeGlobalId(CUSTOMER).id}
    `;
    expect(count).toBe(1n);
  });

  test("operation is idempotent", async () => {
    const result = await execute(schema, TestCreateTemplateConstraintDocument, {
      template: TEMPLATE.id,
      location: LOCATION.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data?.createTemplateConstraint).toMatchObject({
      constraint: {
        id: expect.any(String),
      },
      diagnostics: [],
      instantiations: [],
    });

    const [{ count }] = await sql`
      select count(*)
      from public.worktemplateconstraint
      where worktemplateconstraintcustomeruuid = ${decodeGlobalId(CUSTOMER).id}
    `;
    expect(count).toBe(1n);
  });

  test("request eager instantiation", async () => {
    const result = await execute(schema, TestCreateTemplateConstraintDocument, {
      template: TEMPLATE.id,
      location: LOCATION.id,
      options: {
        instantiate: {},
      },
    });
    expect(
      result.data?.createTemplateConstraint?.instantiations,
    ).toMatchSnapshot();
  });

  test("request eager instantiation with field-level edits", async () => {
    const field = await getFieldByName(TEMPLATE, "Description");
    const result = await execute(schema, TestCreateTemplateConstraintDocument, {
      template: TEMPLATE.id,
      location: LOCATION.id,
      options: {
        instantiate: {
          fields: [
            {
              field: field.id,
              value: {
                string: "First try?",
              },
              valueType: field.valueType,
            },
          ],
        },
      },
    });
    expect(
      result.data?.createTemplateConstraint?.instantiations,
    ).toMatchSnapshot();
  });

  beforeAll(async () => {
    const logs = await sql<{ op: string; id: string }[]>`
      select *
      from
          runtime.create_demo(
              customer_name := 'Frozen Tendy Factory',
              admins := (
                  select array_agg(workeruuid)
                  from public.worker
                  where workeridentityid = ${ctx.auth.userId}
              ),
              modified_by := 895
          )
      ;
    `;
    CUSTOMER = findAndEncode("customer", "organization", logs);
    TEMPLATE = map(
      findAndEncode("next", "worktemplate", logs),
      id => new Task({ id }, ctx),
    );
    LOCATION = map(
      findAndEncode("location", "location", logs),
      id => new Location({ id }, ctx),
    );

    // Delete all constraints. FIXME: don't create a Runtime customer every time.
    await sql`
      delete from public.worktemplateconstraint
      where worktemplateconstraintcustomeruuid = ${decodeGlobalId(CUSTOMER).id}
    `;
  });

  afterAll(async () => {
    const { id } = decodeGlobalId(CUSTOMER);
    process.stdout.write("Cleaning up... ");
    const [row] = await sql<[{ ok: string }]>`
      select runtime.destroy_demo(${id}) as ok;
    `;
    console.log(row.ok);
  });
});
