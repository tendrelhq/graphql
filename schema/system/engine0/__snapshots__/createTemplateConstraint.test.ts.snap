// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createTemplateConstraint request eager instantiation 1`] = `
[
  {
    "__typename": "TaskEdge",
    "node": {
      "__typename": "Task",
      "displayName": {
        "__typename": "DisplayName",
        "name": {
          "__typename": "DynamicString",
          "value": "Idle Time",
        },
      },
      "fields": {
        "__typename": "FieldConnection",
        "edges": [
          {
            "__typename": "FieldEdge",
            "node": {
              "__typename": "Field",
              "name": {
                "__typename": "DisplayName",
                "name": {
                  "__typename": "DynamicString",
                  "value": "Override Start Time",
                },
              },
              "value": {
                "__typename": "TimestampValue",
              },
            },
          },
          {
            "__typename": "FieldEdge",
            "node": {
              "__typename": "Field",
              "name": {
                "__typename": "DisplayName",
                "name": {
                  "__typename": "DynamicString",
                  "value": "Override End Time",
                },
              },
              "value": {
                "__typename": "TimestampValue",
              },
            },
          },
          {
            "__typename": "FieldEdge",
            "node": {
              "__typename": "Field",
              "name": {
                "__typename": "DisplayName",
                "name": {
                  "__typename": "DynamicString",
                  "value": "Description",
                },
              },
              "value": {
                "__typename": "StringValue",
                "string": null,
              },
            },
          },
        ],
      },
    },
  },
]
`;

exports[`createTemplateConstraint request eager instantiation with field-level edits 1`] = `
[
  {
    "__typename": "TaskEdge",
    "node": {
      "__typename": "Task",
      "displayName": {
        "__typename": "DisplayName",
        "name": {
          "__typename": "DynamicString",
          "value": "Idle Time",
        },
      },
      "fields": {
        "__typename": "FieldConnection",
        "edges": [
          {
            "__typename": "FieldEdge",
            "node": {
              "__typename": "Field",
              "name": {
                "__typename": "DisplayName",
                "name": {
                  "__typename": "DynamicString",
                  "value": "Override Start Time",
                },
              },
              "value": {
                "__typename": "TimestampValue",
              },
            },
          },
          {
            "__typename": "FieldEdge",
            "node": {
              "__typename": "Field",
              "name": {
                "__typename": "DisplayName",
                "name": {
                  "__typename": "DynamicString",
                  "value": "Override End Time",
                },
              },
              "value": {
                "__typename": "TimestampValue",
              },
            },
          },
          {
            "__typename": "FieldEdge",
            "node": {
              "__typename": "Field",
              "name": {
                "__typename": "DisplayName",
                "name": {
                  "__typename": "DynamicString",
                  "value": "Description",
                },
              },
              "value": {
                "__typename": "StringValue",
                "string": "First try?",
              },
            },
          },
        ],
      },
    },
  },
]
`;
